Index: ocaml-dbus-0.30/dbus_stubs.c
===================================================================
--- ocaml-dbus-0.30.orig/dbus_stubs.c
+++ ocaml-dbus-0.30/dbus_stubs.c
@@ -331,6 +331,15 @@ value stub_dbus_string_of_error_name(val
 
 /******************** BUS **********************/
 
+static void init_dbus(void)
+{
+	static int done = 0;
+	if (!done) {
+		dbus_threads_init_default();
+		done = 1;
+	}
+}
+
 value stub_dbus_bus_get(value type)
 {
 	CAMLparam1(type);
@@ -338,6 +347,7 @@ value stub_dbus_bus_get(value type)
 	DBusConnection *c_con;
 	DBusError error;
 
+	init_dbus();
 	dbus_error_init(&error);
 	c_con = dbus_bus_get(__bustype_table[Int_val(type)], &error);
 	if (!c_con)
@@ -354,6 +364,7 @@ value stub_dbus_bus_get_private(value ty
 	DBusConnection *c_con;
 	DBusError error;
 
+	init_dbus();
 	dbus_error_init(&error);
 	c_con = dbus_bus_get_private(__bustype_table[Int_val(type)], &error);
 	if (!c_con)
