Index: ocaml-dbus-0.30/dbus_stubs.c
===================================================================
--- ocaml-dbus-0.30.orig/dbus_stubs.c
+++ ocaml-dbus-0.30/dbus_stubs.c
@@ -1143,7 +1143,7 @@ static void message_append_basic(DBusMes
 		}
 	case DBUS_TYPE_OBJECT_PATH:
 	case DBUS_TYPE_STRING: {
-		char *s = strdup(String_val(v));
+		char *s = String_val(v);
 		dbus_message_iter_append_basic(iter, c_type, &s);
 		break;
 		}
