Index: ocaml-dbus-0.30/dbus_stubs.c
===================================================================
--- ocaml-dbus-0.30.orig/dbus_stubs.c
+++ ocaml-dbus-0.30/dbus_stubs.c
@@ -986,7 +986,11 @@ value stub_dbus_message_get_error_name(v
 		error_name = Val_none;
 	else {
 		int index = find_index_string(errname, __error_table);
-		caml_alloc_some(error_name, Val_int(index));
+		if (index >= 0) {
+			caml_alloc_some(error_name, Val_int(index));
+		} else {
+			error_name = Val_none;
+		}
 	}
 	CAMLreturn(error_name);
 }
