Index: xen-4.6.4/tools/firmware/hvmloader/pci.c
===================================================================
--- xen-4.6.4.orig/tools/firmware/hvmloader/pci.c
+++ xen-4.6.4/tools/firmware/hvmloader/pci.c
@@ -160,8 +160,8 @@ void pci_setup(void)
         switch ( class )
         {
         case 0x0300:
-            /* If emulated VGA is found, preserve it as primary VGA. */
-            if ( (vendor_id == 0x1234) && (device_id == 0x1111) )
+            /* If OpenXT emulated VGA is found, preserve it as primary VGA. */
+            if ( (vendor_id == 0x5853) && (device_id == 0xc101) )
             {
                 vga_devfn = devfn;
                 virtual_vga = VGA_std;
